universal-application-tool-0.0.1/app/services/Path.java

```diff
@@ -127,21 +127,27 @@ public Path withoutArrayReference() {
   /**
    * Return the index of the array element this path is referencing.
    *
-   * <p>For paths to non-array elements, {@code IllegalStateException is thrown}.
+   * <p>For paths to non repeated entity collections, or paths to an entire repeated entity
+   * collection without specifying an index, {@code IllegalStateException} is thrown.
    */
   public int arrayIndex() {
     Matcher matcher = ARRAY_INDEX_REGEX.matcher(keyName());
     if (matcher.matches()) {
-      return Integer.valueOf(matcher.group(ARRAY_INDEX_GROUP));
+      try {
+        return Integer.valueOf(matcher.group(ARRAY_INDEX_GROUP));
+      } catch (NumberFormatException e) {
+        throw new IllegalStateException(
+            String.format("This path %s does not reference a repeated entity element.", this), e);
+      }
     }
     throw new IllegalStateException(
-        String.format("This path %s does not reference an array element.", this));
+        String.format("This path %s does not reference a repeated entity collection.", this));
```

Comment:
Suggestion (blocking): the same error is thrown in two branches of this code. Could be re-written more simply as

```java
    Matcher matcher = ARRAY_INDEX_REGEX.matcher(keyName());

    try {
      return Integer.valueOf(matcher.group(ARRAY_INDEX_GROUP));
    } catch (NumberFormatException e) {
      throw new IllegalStateException(
          String.format("This path %s does not reference a repeated entity element.", this), e);
    }
```

Reply:
done! It needs to also catch the `IllegalStateException` for when there is no match at all.